stdlib/Distributed/src/remotecall.jl

```diff
@@ -292,24 +292,29 @@ function serialize(s::ClusterSerializer, rr::AbstractRemoteRef, addclient)
 end
 
 function deserialize(s::ClusterSerializer, t::Type{<:Future})
-    f = deserialize_rr(s,t)
-    Future(f.where, RRID(f.whence, f.id), f.v) # ctor adds to client_refs table
+    f = invoke(deserialize, Tuple{ClusterSerializer, DataType}, s, t)
+    f2 = Future(f.where, RRID(f.whence, f.id), f.v) # ctor adds to client_refs table
+
+    # 1) send_add_client() is not executed when the ref is being serialized
+    #    to where it exists, hence do it here.
+    # 2) If we have recieved a 'fetch'ed Future or if the Future ctor found an
+    #    already 'fetch'ed instance in client_refs (Issue #25847), we should not
+    #    track it in the backing RemoteValue store.
+    if f2.where == myid() && f2.v === nothing
```

Comment:
Could you explain a bit more how this works? It seems like a data race to me --- we `add_client` for the current process only if the future is not done computing yet, and we don't know when that will happen.

Reply:
Given a Future created on node 1 but pointing to data on node 2 (i.e., where=2, whence=1) :

If it is unfetched and sent to node 3, `add_client` is invoked here - https://github.com/JuliaLang/julia/blob/4115c9cb88a9e81faacff4ba41a5d420b3f6cf44/stdlib/Distributed/src/remotecall.jl#L287-L290
(Note that the `if addclient` is checking if the Future is unfetched, i.e. same as `if f.v === nothing) 

If it has been fetched and is being sent to 3, no `add_client` is performed since it is assumed that the fetch call would have removed the reference to the backing store and it may been collected on the remote node.
The del_client here - https://github.com/JuliaLang/julia/blob/4115c9cb88a9e81faacff4ba41a5d420b3f6cf44/stdlib/Distributed/src/remotecall.jl#L494

Now, add_client as a message is not sent if the Future is being sent to the "where" node, 2 in this case, as the add_client will be performed during the deserialization process.

The bug in this case manifested during the second time the Future was serialized to its "where" node. The first time it was sent, the `Future` constructor did not find an entry in `client_refs`, created one and setup a finalizer - https://github.com/JuliaLang/julia/blob/4115c9cb88a9e81faacff4ba41a5d420b3f6cf44/stdlib/Distributed/src/remotecall.jl#L64-L79 . In the bug report posted, the remote function invoked with this Future did a `fetch`. This resulted in a) the "whence" id, i.e. 1  in this case, being removed from PGRP.refs[rrid].clientset (as `fetch` invokes a del_client the first time it is invoked on a Future), and b) Future.v being populated with the computation value.
Now the second time this same Future was sent to 2, the deserialization/Future constructor code immediately returned a cached instance (from `client_refs`) and an add_client was done. The remote function invoked `fetch`, which again directly returned `Future.v` (did not execute a `del_client` as was done for the first time `fetch` was called). This extra `add_client` was the source of the leak as the finalizer code does NOT execute a `del_client` on fetched Futures. See here - 
https://github.com/JuliaLang/julia/blob/4115c9cb88a9e81faacff4ba41a5d420b3f6cf44/stdlib/Distributed/src/remotecall.jl#L93

The fix avoids doing an add_client if a Future has been fetched in both cases (i.e., destination node same as "where" node and when it is different).

Reply:
Great explanation, thank you!